File: python/234.palindrome-linked-list.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # if list is empty or has only one node
        if not head or not head.next:
            return True

        # find the middle, and reverse left part
        slow, fast, prev = head, head, None
        while slow and fast and fast.next:
            tmp = slow

            slow = slow.next
            fast = fast.next.next

            tmp.next = prev
            prev = tmp

        # if nodes count is odd, skip the middle node
        if fast and slow:
            slow = slow.next

        # compare
        while prev and slow and prev.val == slow.val:
            slow = slow.next
            prev = prev.next

        # if reached last, return True
        return not prev

File: golang/234.palindrome-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func isPalindrome(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return true
	}

	fast, slow := head, head
	var prev, tmp *ListNode
	for fast != nil && fast.Next != nil {
		tmp = slow
		slow = slow.Next
		fast = fast.Next.Next

		tmp.Next = prev
		prev = tmp
	}

	if fast != nil {
		slow = slow.Next
	}

	for prev != nil && prev.Val == slow.Val {
		prev = prev.Next
		slow = slow.Next
	}

	if slow == nil && prev == nil {
		return true
	}

	return false
}
